Approved/remove-new-array.md

```diff
@@ -0,0 +1,84 @@
+---
+title: Remove default-initialized new array expressions
+description: The `new Type[N]` array expression is deprecated and removed from Q#.
+author: Sarah Marshall
+date: October 28, 2020
+---
+
+# Proposal
+
+The `new Type[N]` expression, which creates an array of `N` values using the default value of `Type`, is deprecated and removed from Q#.
+
+# Justification
+
+The existence of `new Type[N]` has negative effects on the language.
+
+The most important negative effect is the assumption that every type has a default value.
+This is not a reasonable assumption, because it is not possible to define useful default values of types like `Qubit` and `a -> b`.
+Their current default values are invalid and will trigger an error if they are used.
+But the syntax is the same both for types with valid default values, and those without, which is misleading: while `new Int[3]` is a well-defined value whose items can be used safely, `new Qubit[3]` creates a potential trap that will crash the program if any item in the array is used before being set to a valid value first.
+
+This assumption is also invalid in the presence of uninhabited types such as `Void`.
+If Q# requires that every type has a default value, then uninhabited types are not possible to express properly.
+This can even cause subtle bugs in the soundness of a type system, as demonstrated by [this bug in Java's generics](https://hackernoon.com/java-is-unsound-28c84cb2b3f) that happened because of the existence of a value (specifically `null`) for a type that should have been uninhabited.
+
+Finally, the use of the `new` keyword is inconsistent with other ways to create values in Q#.
+`new` is required only when creating a default-initialized array, not when using a square bracket array literal or when creating values of other types.
+
+# Description
+
+## Current Status
+
+Q# currently supports this syntax:
+
+```qsharp
+// Create an empty array of type Int[].
+let empty = new Int[0];
+
+// Create an array of length 10 of type Bool[] where every value is false.
+let bools = new Bool[10];
+```
+
+## Proposed Modification
+
+The syntax above will be removed.
+The same functionality can be accomplished with functions in the standard library:
+
+```qsharp
+// Create an empty arra of type Int[].
+let empty = EmptyArray<Int>();
+
+// Create an array of length 10 of type Bool[] where every value is false.
+let bools = ConstantArray(10, false);
+```
+
+# Implementation
```

Comment:
I believe EmptyArray and ConstantArray right now are not intrinsic functions. We would need either a special syntax or an intrinsic function that connects to the corresponding runtime functionality.

Reply:
Quick side note, but while ConstantArray is implemented in Q#, EmptyArray is already intrinsic in the runtime (https://github.com/microsoft/qsharp-runtime/blob/a18ecf4eb1a6cad4800c6dfe4a5ce96e362d807c/src/Simulation/QsharpCore/Arrays/Empty.cs).

Reply:
Added some more about this. Technically, `ConstantArray` doesn't need to be intrinsic, but it could be more efficient if it was, so I added a note that we should do that.

Reply:
Irrespective of how this proposal goes more generally, I agree with moving `ConstantArray` to being intrinsic as a performance consideration. If possible, it may be good to keep the pure-Q# fallback to make it easier for custom simulators by reducing how much they have to implement.

Reply:
I forgot that you can still have a Q# fallback for a callable that has a native implementation. I clarified this in the proposal.

Reply:
I don't think having it be intrinsic necessarily impose an additional burden on the runtime beyond what already needs to be supported; the ConstantArray can be recognized and replaced with whatever construct the runtime supports prior to linking. 
@alan-geller for input on whether we want to require such a function being supported in QIR runtime regardless. 
@swernli for guidance on how it best fits with the recent C# runtime restructuring (I believe the way it aligns with what else we have would be to have it be an intrinsic function in the Q# foundation).